### PR TITLE
Implementation of az module for alertsmanagement

### PR DESCRIPTION
**Pull request by mahakjain314**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by mahakjain314** [comment]
@zikalino Need assistance on fixing the below style check -
R0915: Too many statements (106/50) (too-many-statements)

**Comment by zikalino** [comment]
@mahakjain314
Please refer to here how to suppress: https://github.com/Azure/azure-cli/blob/2a5540e94eb4160d87d883980fc9c906f7fc0789/src/azure-cli/azure/cli/command_modules/appservice/_params.py#L34

Also, yes, you need to retarget this PR to *dev*.
In a meanwhile I will continue the review.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        c.argument('description_condition', options_list=['--description-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Contains:Test Alert')
+        c.argument('alert_context_condition', options_list=['--alert-context-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>} For eg. Contains:smartgroups')
+        c.argument('action_rule_type', options_list=['--action-rule-type'], required=True, help='Action rule Type')
+        c.argument('recurrence_type', options_list=['--recurrence-type'], required=False,
+                   help='Specifies the duration when the suppression should be applied')
+        c.argument('suppression_start_time', options_list=['--suppression-start-time'], required=False,
+                   help='Suppression Start Time. Format 12/09/2018 06:00:00' +
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Could we use ISO8601 here as well? It looks like it's some kind of custom format.

**Comment by zikalino** [comment]
Pls add entry on the top of HISTORY.rst: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/HISTORY.rst

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        c.argument('alert_ruleId_condition', options_list=['--alert-ruleId-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Equals:ARM_ID_1,ARM_ID_2')
+        c.argument('description_condition', options_list=['--description-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Contains:Test Alert')
+        c.argument('alert_context_condition', options_list=['--alert-context-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>} For eg. Contains:smartgroups')
+        c.argument('action_rule_type', options_list=['--action-rule-type'], required=True, help='Action rule Type')
+        c.argument('recurrence_type', options_list=['--recurrence-type'], required=False,
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: shouldn't this be enum?
I can't find explicitly defined values

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        c.argument('severity_condition', options_list=['--severity-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>} For eg. Equals:Sev0,Sev1')
+        c.argument('monitor_service_condition', options_list=['--monitor-service-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Equals:Platform,Log Analytics')
+        c.argument('monitor_condition', monitor_condition_arg_type, required=False)
+        c.argument('target_resource_type_condition', options_list=['--target-resource-type-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Contains:Virtual Machines,Storage Account')
+        c.argument('alert_ruleId_condition', options_list=['--alert-ruleId-condition'], required=False,
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: shouldn't it be alert_rule_id_condition?

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Equals:ARM_ID_1,ARM_ID_2')
+        c.argument('description_condition', options_list=['--description-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>}' +
+                   ' For eg. Contains:Test Alert')
+        c.argument('alert_context_condition', options_list=['--alert-context-condition'], required=False,
+                   help='Expected format - {<operation>:<comma separated list of values>} For eg. Contains:smartgroups')
+        c.argument('action_rule_type', options_list=['--action-rule-type'], required=True, help='Action rule Type')
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Is action_rule_type an enum?
I found following values: ActionGroup, Suppression, Diagnostic

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_help.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+  - name: --recurrence-values
+    type: string
+    short-summary: Reccurent values, if applicable. In case of Weekly - 1,3,5 In case of Monthly - 16,24,28
+  - name: --action-group-id
+    type: string
+    short-summary: Action Group Id which is to be notified
+examples:
+  - name: Set action-rule on various parameters.
+    text: >
+        az alertsmanagement action-rule set --resource-group-name "test-rg" --name "Test-AR" --scope "/subscriptions/dd91de05-d791-4ceb-b6dc-988682dc7d72/resourceGroups/alertslab","/subscriptions/dd91de05-d791-4ceb-b6dc-988682dc7d72/resourceGroups/Test-VMs" --severity-condition "Equals:Sev0,Sev1" --monitor-condition "NotEquals:Resolved" --description "Test description" --status "Enabled" --action-rule-type "Suppression" --recurrence-type "Weekly" --suppression-start-time "06/26/2018 06:00:00" --suppression-end-time "07/27/2018 06:00:00" --recurrence-values 1,4,6
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_help.py]: Usually in examples we use "MyResourceGroup" as resource group name. Also alert name coulld be "MyAlert".
Also it's better not to include real subscription id "dd91de05-d791-4ceb-b6dc-988682dc7d72"

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        c.argument('recurrence_type', options_list=['--recurrence-type'], required=False,
+                   help='Specifies the duration when the suppression should be applied')
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Does this parameter apply only to **--action-rule-type Suppression**?

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 3 above)
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: could there be more than one condition?

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        c.argument('description', options_list=['--description'], required=False,
+                   help='Gets all actions rules in a subscription filter by description')
+        c.argument('alert_rule', alert_rule_arg_type, required=False)
+        c.argument('action_group', options_list=['--action-group'], required=False,
+                   help='Gets all actions rules in a subscription filter by action group')
+        c.argument('name', options_list=['--name'], required=False,
+                   help='Gets all actions rules in a subscription filter by action rule name')
+
+    with self.argument_context('alertsmanagement action-rule show') as c:
+        c.argument('resource_group_name', resource_group_name_arg_type, required=True)
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: The argument should be 'resource_group'. It should look as follows in help:

    --resource-group -g         [Required] : Name of resource group. You can configure the default
                                             group using `az configure --defaults group=<name>`.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+                                          help='Sorting order. Supported values – asc, desc',
+                                          metavar='SORT_ORDER')
+    time_range_arg_type = CLIArgumentType(options_list=['--time-range'],
+                                          help='Supported time range values – 1h, 1d, 7d, 30d (Default is 1d)',
+                                          metavar='TIME_RANGE')
+    custom_time_range_arg_type = CLIArgumentType(options_list=['--custom-time-range'],
+                                                 help='Supported format - <start-time>/<end-time>' +
+                                                 ' where time is in ISO-8601 format',
+                                                 metavar='CUSTOM_TIME_RANGE')
+    resource_group_name_arg_type = CLIArgumentType(options_list=['--resource-group-name'],
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: don't define resource_group_name_arg_type here, just import default one:

from azure.cli.core.commands.parameters import (
   ….
    resource_group_name_type,
   …..

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    monitor_service_arg_type = CLIArgumentType(options_list=['--monitor-service'], help='Filter by target resource',
+                                               metavar='MONITOR_SERVICE')
+    monitor_condition_arg_type = CLIArgumentType(options_list=['--monitor-condition'],
+                                                 help='Filter by monitor condition. Supported values – Fired, Resolved',
+                                                 metavar='MONITOR_CONDITION')
+    severity_arg_type = CLIArgumentType(options_list=['--severity'],
+                                        help='Filter by severity. Supported values – Sev0, Sev1, Sev2, Sev3, Sev4',
+                                        metavar='SEVERITY')
+    state_arg_type = CLIArgumentType(options_list=['--state'],
+                                     help='Filter by alert state. Supported values - New, Acknowledged, Closed',
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: could that be defined as enum?
If you define enum argument type, values would be included automatically in help.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    target_resource_group_arg_type = CLIArgumentType(options_list=['--target-resource-group'],
+                                                     help='Filter by resource group',
+                                                     metavar='TARGET_RESOURCE_GROUP')
+    monitor_service_arg_type = CLIArgumentType(options_list=['--monitor-service'], help='Filter by target resource',
+                                               metavar='MONITOR_SERVICE')
+    monitor_condition_arg_type = CLIArgumentType(options_list=['--monitor-condition'],
+                                                 help='Filter by monitor condition. Supported values – Fired, Resolved',
+                                                 metavar='MONITOR_CONDITION')
+    severity_arg_type = CLIArgumentType(options_list=['--severity'],
+                                        help='Filter by severity. Supported values – Sev0, Sev1, Sev2, Sev3, Sev4',
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: enum here as well?

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    target_resource_type_arg_type = CLIArgumentType(options_list=['--target-resource-type'],
+                                                    help='Filter by target resource type',
+                                                    metavar='TARGET_RESOURCE_TYPE')
+    target_resource_group_arg_type = CLIArgumentType(options_list=['--target-resource-group'],
+                                                     help='Filter by resource group',
+                                                     metavar='TARGET_RESOURCE_GROUP')
+    monitor_service_arg_type = CLIArgumentType(options_list=['--monitor-service'], help='Filter by target resource',
+                                               metavar='MONITOR_SERVICE')
+    monitor_condition_arg_type = CLIArgumentType(options_list=['--monitor-condition'],
+                                                 help='Filter by monitor condition. Supported values – Fired, Resolved',
```
Comment (zikalino) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: enum?

File: azure-cli2017.pyproj [hunk 12]
```diff
@@ -11,7 +11,7 @@
     <OutputPath>.</OutputPath>
     <ProjectTypeGuids>{888888a0-9f3d-457c-b088-3a5042f75d52}</ProjectTypeGuids>
     <LaunchProvider>Standard Python launcher</LaunchProvider>
-    <InterpreterId>MSBuild|env|$(MSBuildProjectFullPath)</InterpreterId>
+    <InterpreterId>MSBuild|env1|$(MSBuildProjectFullPath)</InterpreterId>
```
Comment (zikalino) [inline azure-cli2017.pyproj]: i think that's probably local change, let's not merge it

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 6 above)
Comment (Rajwanshi) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Yes reccurence_type only applies to suppression.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 1 above)
Comment (Rajwanshi) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: This custom format is used as this is in accordance with corresponding API calls.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/commands.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        exception_handler=alertsmanagement_exception_handler
+    )
+
+    alertsmanagement_action_rule_util = CliCommandType(
+        operations_tmpl='azure.mgmt.alertsmanagement.operations#ActionRulesOperations.{}',
+        client_factory=action_rules_mgmt_client_factory,
+        exception_handler=alertsmanagement_exception_handler
+    )
+
+    with self.command_group('alertsmanagement alert', alertsmanagement_alert_util,
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/alertsmanagement/commands.py]: Hello. Is this module related with monitor? Why don't we include this module under "az monitor"?

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 7 above)
Comment (Rajwanshi) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Required changes have been done. Thanks!

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 8 above)
Comment (Rajwanshi) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Required changes have been done. Thanks!

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 3 above)
Comment (Rajwanshi) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: Conditions are mutually exclusive and there can be more than one condition

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 1 above)
Comment (zikalino) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: well, I think we should still fix it here.
Azure CLI shouldn't expose inconsistencies like that.

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py (hunk 6 above)
Comment (zikalino) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/_params.py]: yes, so the documentation should be more clear about it

File: src/azure-cli/azure/cli/command_modules/alertsmanagement/commands.py (hunk 13 above)
Comment (zikalino) [reply src/azure-cli/azure/cli/command_modules/alertsmanagement/commands.py]: @MyronFanQiu, yes that should be definitely considered

**Comment by qwordy** [comment]
@mahakjain314, @Rajwanshi, are you still working on this PR?
